scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroBucketMetadata.java

```diff
@@ -25,30 +25,29 @@
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.annotation.JsonProperty;
-import java.util.Arrays;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import org.apache.avro.Schema;
-import org.apache.avro.generic.GenericRecord;
-import org.apache.avro.reflect.ReflectData;
+import org.apache.avro.generic.IndexedRecord;
 import org.apache.beam.sdk.coders.CannotProvideCoderException;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.transforms.display.DisplayData;
 import org.apache.beam.sdk.transforms.display.DisplayData.Builder;
 
 /**
- * {@link org.apache.beam.sdk.extensions.smb.BucketMetadata} for Avro {@link GenericRecord} records.
+ * {@link org.apache.beam.sdk.extensions.smb.BucketMetadata} for Avro {@link IndexedRecord} records.
  */
-public class AvroBucketMetadata<K1, K2, V extends GenericRecord> extends BucketMetadata<K1, K2, V> {
+public class AvroBucketMetadata<K1, K2, V extends IndexedRecord> extends BucketMetadata<K1, K2, V> {
 
   @JsonProperty private final String keyField;
 
   @JsonProperty
   @JsonInclude(JsonInclude.Include.NON_NULL)
   private final String keyFieldSecondary;
 
-  @JsonIgnore private final String[] keyPath;
-  @JsonIgnore private final String[] keyPathSecondary;
+  @JsonIgnore private final AtomicReference<int[]> keyPath = new AtomicReference<>();
+  @JsonIgnore private final AtomicReference<int[]> keyPathSecondary = new AtomicReference<>();
```

Comment:
We now have to delay the construction for those, as the schema is required to get access to the field index instead of the field name